### PR TITLE
Support for Kubernetes v1.19

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area open-source usability
/kind enhancement
/priority normal

**What this PR does / why we need it**:
Support for Kubernetes v1.19

**Which issue(s) this PR fixes**:
Part of #2629

**Special notes for your reviewer**:
* I didn't vendor the `kubernetes-1.19` sources yet due to the ongoing work of #2489 and the dependency of k-sigs/controller-runtime to 1.18 as of now.
* ⚠️ This PR only contains the Gardener part for supporting 1.19 - there will/must be follow-up PRs for each extension repository where individual support gets added.
* ⚠️ Due to the fact that the Kubernetes community no longer builds `hyperkube` images on their own we have created [gardener/hyperkube](https://github.com/gardener/hyperkube). It produces Docker images containing only the `kubelet` and `kubectl` binaries which are used to bootstrap the shoot worker nodes. Please note that this means that new Kubernetes versions are now only supported by Gardener if there is a corresponding release on the [gardener/hyperkube](https://github.com/gardener/hyperkube) repository.

**Release note**:

```noteworthy operator
Gardener can now support shoot clusters with Kubernetes version 1.19. In order to allow creation/update of 1.19 clusters you will have to update the version of your provider extension(s) to a version that supports 1.19 as well. Please consult the respective releases and notes in the provider extension's repository.
```
```noteworthy operator
Due to the fact that the Kubernetes community no longer builds `hyperkube` images on their own we have created [gardener/hyperkube](https://github.com/gardener/hyperkube). It produces Docker images containing only the `kubelet` and `kubectl` binaries which are used to bootstrap the shoot worker nodes. Please note that this means that new Kubernetes versions are now only supported by Gardener if there is a corresponding release on the [gardener/hyperkube](https://github.com/gardener/hyperkube) repository.
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -32,14 +32,14 @@ Please find more information regarding the concepts and a detailed description o
...
-| **Alicloud** | [![Gardener v1.18 Conformance Tests](https://testgrid.k8s.io/q/summary/conformance-gardener/Gardener,%20v1.18%20Alibaba%20Cloud/tests_status?style=svg)](https://testgrid.k8s.io/conformance-gardener#Gardener,%20v1.18%20Alibaba%20Cloud) | [![Gardener v1.17 Conformance Tests](https://testgrid.k8s.io/q/summary/conformance-gardener/Gardener,%20v1.17%20Alibaba%20Cloud/tests_status?style=svg)](https://testgrid.k8s.io/conformance-gardener#Gardener,%20v1.17%20Alibaba%20Cloud) | [![Gardener v1.16 Conformance Tests](https://testgrid.k8s.io/q/summary/conformance-gardener/Gardener,%20v1.16%20Alibaba%20Cloud/tests_status?style=svg)](https://testgrid.k8s.io/conformance-gardener#Gardener,%20v1.16%20Alibaba%20Cloud) | [2] | [2] | [2] | N/A | N/A | N/A
-| **Packet** | N/A | N/A | N/A | N/A | N/A | N/A | N/A | N/A | N/A |
+| Provider/K8s | v1.19 | v1.18 | v1.17 | v1.16 | v1.15 | v1.14 |  v1.13 |  v1.12 |  v1.11 |  v1.10 |
```
Comment (rfranzke) [inline README.md:35]: cc @schrodit @dguendisch - looking forward to get some conformance test results :)

File: charts/utils-templates/templates/_versions.tpl [hunk 2]
```diff
@@ -81,7 +81,9 @@ policy/v1beta1
 {{- end -}}
```
Comment (vpnachev) [inline charts/utils-templates/templates/_versions.tpl:73]: The comment is about line 7.
`kubescheduler.config.k8s.io/v1alpha1` has been removed in v1.19+
ref: https://github.com/kubernetes/kubernetes/pull/89298

The available APIs are  kubescheduler.config.k8s.io/v1alpha2 and kubescheduler.config.k8s.io/v1beta1, but we are not using them anywhere.

**Review by vpnachev — COMMENTED** [review]
I have quickly go through this PR and the release notes and have 2 proposals.

`kubelet_running_pod_count` is replaced by `kubelet_running_pods` https://github.com/gardener/gardener/blob/master/charts/seed-monitoring/charts/core/charts/prometheus/templates/config.yaml#L185 ref: https://github.com/kubernetes/kubernetes/pull/92407

**Comment by ialidzhikov** [comment]
I have also quickly go through.

I was able to create successfully v1.19.0 Shoot.

```
$ k get no
NAME                                       STATUS   ROLES    AGE   VERSION
ip-10-222-0-4.eu-west-1.compute.internal   Ready    <none>   12m   v1.19.0
```

For now I spotted issues with metrics-server:
```
$ k -n kube-system top po
W0901 00:06:45.907731   85588 top_pod.go:274] Metrics not available for pod kube-system/calico-kube-controllers-68657c7d94-kkmgb, age: 15m44.907719s
error: Metrics not available for pod kube-system/calico-kube-controllers-68657c7d94-kkmgb, age: 15m44.907719s
```

Is it also reproducible for you?

**Comment by rfranzke** [comment]
Nice catch @ialidzhikov - thank you. It seems there is an issue with Docker 18 and Kubernetes 1.19, see https://github.com/kubernetes/kubernetes/issues/94281. Can you try upgrading to an OS that uses Docker 19?

File: pkg/apis/core/validation/shoot_test.go [hunk 3]
```diff
@@ -1329,6 +1329,43 @@ var _ = Describe("Shoot Validation Tests", func() {
 			})
 		})
 
+		Context("basic authentication", func() {
+			BeforeEach(func() {
+				shoot.Spec.Kubernetes.KubeControllerManager.HorizontalPodAutoscalerConfig.DownscaleDelay = nil
+				shoot.Spec.Kubernetes.KubeControllerManager.HorizontalPodAutoscalerConfig.UpscaleDelay = nil
+			})
```
Comment (timebertt) [inline pkg/apis/core/validation/shoot_test.go:1336]: What is the purpose of this `BeforeEach` block?

File: pkg/operation/botanist/cleanup.go [hunk 4]
```diff
@@ -202,7 +202,7 @@ func (b *Botanist) CleanExtendedAPIs(ctx context.Context) error {
 	)
 
 	return flow.Parallel(
-		cleanResourceFn(defaultOps, c, &apiregistrationv1beta1.APIServiceList{}, APIServiceCleanOption, GracePeriodFiveMinutes, FinalizeAfterOneHour),
+		cleanResourceFn(defaultOps, c, &apiregistrationv1.APIServiceList{}, APIServiceCleanOption, GracePeriodFiveMinutes, FinalizeAfterOneHour),
```
Comment (timebertt) [inline pkg/operation/botanist/cleanup.go:205]: This still works on older clusters which don't have the `apiregistration/v1` API, right?

File: charts/seed-operatingsystemconfig/original/templates/kubelet/_kubelet-monitor.service [hunk 5]
```diff
@@ -13,8 +13,10 @@
     EnvironmentFile=/etc/environment
     {{- if semverCompare "< 1.17" .Values.kubernetes.version }}
     ExecStartPre=/usr/bin/docker run --rm -v /opt/bin:/opt/bin:rw {{ required "images.hyperkube is required" .Values.images.hyperkube }} /bin/sh -c "cp /usr/local/bin/kubectl /opt/bin"
-    {{- else }}
+    {{- else if semverCompare "< 1.19" .Values.kubernetes.version }}
     ExecStartPre=/usr/bin/docker run --rm -v /opt/bin:/opt/bin:rw --entrypoint /bin/sh {{ required "images.hyperkube is required" .Values.images.hyperkube }} -c "cp /usr/local/bin/kubectl /opt/bin"
+    {{- else }}
+    ExecStartPre=sh -c "ID=\"$(/usr/bin/docker run --rm -d -v /opt/bin:/opt/bin:rw {{ required "images.hyperkube is required" .Values.images.hyperkube }})\"; /usr/bin/docker cp \"$ID\":/kubectl /opt/bin; docker stop \"$ID\"; chmod +x /opt/bin/kubectl"
```
Comment (timebertt) [inline charts/seed-operatingsystemconfig/original/templates/kubelet/_kubelet-monitor.service]: Any specific reason, why we are using two different strategies to copy the binary for different k8s versions now (`cp` in container / `docker cp`)?

**Review by timebertt — COMMENTED** [review]
Looks pretty good!
Just a few questions :)

File: charts/seed-operatingsystemconfig/original/templates/kubelet/_kubelet-monitor.service (hunk 5 above)
Comment (rfranzke) [reply charts/seed-operatingsystemconfig/original/templates/kubelet/_kubelet-monitor.service]: The gardener/hyperkube image is based on `scratch` and does not have anything other than the binaries inside, so especially it doesn't have a shell/bash. Hence, we have to use `docker cp` now.

File: pkg/apis/core/validation/shoot_test.go (hunk 3 above)
Comment (rfranzke) [reply pkg/apis/core/validation/shoot_test.go:1336]: These fields are not valid for Kubernetes 1.12+, but we are having tests for 1.18 and 1.19 below. `ValidateShoot` would show errors for the `{Down,Up}scaleDelay` fields, so I'm setting them to `nil` to let the test focus on the basic authentication part.

File: pkg/operation/botanist/cleanup.go (hunk 4 above)
Comment (rfranzke) [reply pkg/operation/botanist/cleanup.go:205]: Yes, it's also defined in the 1.10 APIs (which is the lowest supported Kubernetes version for shoots by Gardener), see https://github.com/kubernetes/kube-aggregator/blob/release-1.10/pkg/apis/apiregistration/v1/types.go#L24

File: pkg/apis/core/validation/shoot_test.go (hunk 3 above)
Comment (timebertt) [reply pkg/apis/core/validation/shoot_test.go:1336]: Got it, thanks.

File: charts/seed-operatingsystemconfig/original/templates/kubelet/_kubelet-monitor.service (hunk 5 above)
Comment (timebertt) [reply charts/seed-operatingsystemconfig/original/templates/kubelet/_kubelet-monitor.service]: Got it, makes sense.
Can we do it always via `docker cp` then, to keep it consistent?

**Comment by vpnachev** [comment]
> Nice catch @ialidzhikov - thank you. It seems there is an issue with Docker 18 and Kubernetes 1.19, see [kubernetes/kubernetes#94281](https://github.com/kubernetes/kubernetes/issues/94281). Can you try upgrading to an OS that uses Docker 19?

I tried and it is still failing. The errors of the metrics server are:
```
E0901 09:09:17.599689       1 reststorage.go:160] unable to fetch pod metrics for pod kube-system/coredns-97dfb4f59-5vmpp: no metrics known for pod
E0901 09:09:17.599706       1 reststorage.go:160] unable to fetch pod metrics for pod kube-system/coredns-97dfb4f59-bhkpf: no metrics known for pod
I0901 09:09:24.301617       1 manager.go:95] Scraping metrics from 1 sources
I0901 09:09:24.306743       1 manager.go:120] Querying source: kubelet_summary:ip-10-250-1-78.eu-west-1.compute.internal
E0901 09:09:38.561769       1 reststorage.go:135] unable to fetch node metrics for node "ip-10-250-1-78.eu-west-1.compute.internal": no metrics known for node
E0901 09:09:47.595897       1 reststorage.go:160] unable to fetch pod metrics for pod kube-system/coredns-97dfb4f59-bhkpf: no metrics known for pod
E0901 09:09:47.595913       1 reststorage.go:160] unable to fetch pod metrics for pod kube-system/coredns-97dfb4f59-5vmpp: no metrics known for pod
I0901 09:09:54.307061       1 manager.go:148] ScrapeMetrics: time: 30.005413203s, nodes: 0, pods: 0
E0901 09:09:54.307088       1 manager.go:111] unable to fully collect metrics: unable to fully scrape metrics from source kubelet_summary:ip-10-250-1-78.eu-west-1.compute.internal: unable to fetch metrics from Kubelet ip-10-250-1-78.eu-west-1.compute.internal (ip-10-250-1-78.eu-west-1.compute.internal): Get https://ip-10-250-1-78.eu-west-1.compute.internal:10250/stats/summary?only_cpu_and_memory=true: dial tcp: i/o timeout
```

The node is running with docker 19.3.7
```
NAME                                        STATUS   ROLES    AGE    VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE       KERNEL-VERSION        CONTAINER-RUNTIME
ip-10-250-1-78.eu-west-1.compute.internal   Ready    <none>   146m   v1.19.0   10.250.1.78   <none>        Garden Linux   5.4.0-4-cloud-amd64   docker://19.3.7
```

**Comment by rfranzke** [comment]
@vpnachev I suggest to put more information on the upstream issue linked above. I guess it's not Gardener but Kubernetes-related.

**Comment by ialidzhikov** [comment]
@vpnachev , it worked for me with gardenlinux `27.1.0`:
```
$ k get no -o wide
NAME                                        STATUS   ROLES    AGE    VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE       KERNEL-VERSION        CONTAINER-RUNTIME
ip-10-222-0-29.eu-west-1.compute.internal   Ready    <none>   3m7s   v1.19.0   10.222.0.29   <none>        Garden Linux   5.4.0-4-cloud-amd64   docker://19.3.7
ip-10-222-0-48.eu-west-1.compute.internal   Ready    <none>   94s    v1.19.0   10.222.0.48   <none>        Garden Linux   5.4.0-4-cloud-amd64   docker://19.3.7
$ k -n kube-system top po
NAME                                                  CPU(cores)   MEMORY(bytes)
calico-kube-controllers-68657c7d94-jb69x              2m           11Mi
calico-node-8s826                                     67m          28Mi
calico-node-jhbxk                                     0m           12Mi
calico-node-vertical-autoscaler-5788bdd9cb-ksmjp      1m           9Mi
calico-typha-deploy-7bc68bcd86-f4chg                  5m           16Mi
calico-typha-horizontal-autoscaler-86cfb97885-t7shd   1m           9Mi
calico-typha-vertical-autoscaler-55b85db9c9-r7jhc     1m           9Mi
coredns-97dfb4f59-jc7g5                               0m           13Mi
coredns-97dfb4f59-qjbg4                               2m           13Mi
csi-driver-node-t6dwn                                 1m           23Mi
csi-driver-node-x695w                                 1m           21Mi
kube-proxy-68rx6                                      1m           18Mi
kube-proxy-b9zq8                                      1m           17Mi
metrics-server-68498c577c-9d674                       2m           16Mi
node-problem-detector-bd966                           0m           7Mi
node-problem-detector-n9mzk                           1m           15Mi
vpn-shoot-7666fbf977-d4jgc                            1m           3Mi
```

File: charts/seed-operatingsystemconfig/original/templates/kubelet/_kubelet-monitor.service (hunk 5 above)
Comment (ialidzhikov) [inline charts/seed-operatingsystemconfig/original/templates/kubelet/_kubelet-monitor.service]: I tried this PR also with suse-chost images and machines were failing to join. 

```
# systemctl status kubelet
● kubelet.service - kubelet daemon
   Loaded: error (Reason: Exec format error)
   Active: inactive (dead)
     Docs: https://kubernetes.io/docs/admin/kubelet

# journalctl -u kubelet

Sep 01 11:55:54 ip-10-222-0-12 systemd[1]: /etc/systemd/system/kubelet.service:15: Executable path is not absolute: sh -c "ID=\"$(/usr/bin/docker run --rm -d -v /opt/bin:/opt/bin:rw eu.gcr.io/gardener-project/hyperkube:v1.19.0)\"; /usr/bin/docker cp \"$ID\":/kubelet /opt/bin; docker stop \"$ID\"; chmod +x /opt/bin/kubelet"
```

Comment (rfranzke) [reply charts/seed-operatingsystemconfig/original/templates/kubelet/_kubelet-monitor.service]: Let's keep it like that, please. I tried to harmonize it, but it didn't work well. It's probably cleaner and easier to maintain to have the three cases for the three different hyperkube images we use.

**Review by ialidzhikov — COMMENTED** [review]
Generally lgtm.
I played with the PR in more extensive manner and it looks promising. I spotted 1 more issue - trying to switch v1.19.0 kube-proxy mode from `IPVS` to `IPTables` failed with:

```
$ k -n kube-system get po -l role=proxy
NAME               READY   STATUS       RESTARTS   AGE
kube-proxy-8dtxn   0/1     Init:Error   6          5m59s
kube-proxy-l5tq9   1/1     Running      0          13m

$ k -n kube-system logs kube-proxy-8dtxn cleanup

W0901 10:59:41.973259       9 server.go:226] WARNING: all flags other than --config, --write-config-to, and --cleanup are deprecated. Please begin using a config file ASAP.
I0901 10:59:41.973396       9 feature_gate.go:243] feature gates: &{map[]}
I0901 10:59:41.973562       9 feature_gate.go:243] feature gates: &{map[]}
I0901 10:59:41.973598       9 server.go:555] Neither kubeconfig file nor master URL was specified. Falling back to in-cluster config.
F0901 10:59:41.973668       9 server.go:495] open /var/run/secrets/kubernetes.io/serviceaccount/token: no such file or directory
```

I am completely fine if you prefer to put this out of the scope of this PR.

File: charts/seed-operatingsystemconfig/original/templates/kubelet/_kubelet-monitor.service (hunk 5 above)
Comment (rfranzke) [reply charts/seed-operatingsystemconfig/original/templates/kubelet/_kubelet-monitor.service]: I changed it to `/usr/bin/env sh`, thanks. Can you confirm it works now?

**Comment by rfranzke** [comment]
@ialidzhikov I have pushed https://github.com/gardener/gardener/pull/2799/commits/8f994259fcc2575bf2bda0e3cca7919d4729c8a8, WDYT?

**Comment by ialidzhikov** [comment]
> @ialidzhikov I have pushed [8f99425](https://github.com/gardener/gardener/commit/8f994259fcc2575bf2bda0e3cca7919d4729c8a8), WDYT?

Works now. Thanks.

File: charts/seed-operatingsystemconfig/original/templates/kubelet/_kubelet-monitor.service (hunk 5 above)
Comment (ialidzhikov) [reply charts/seed-operatingsystemconfig/original/templates/kubelet/_kubelet-monitor.service]: Yep, works now.

**Review by ialidzhikov — APPROVED** [review]
/lgtm

**Review by vpnachev — APPROVED** [review]
/lgtm 

My troubles were caused by the `NodeLocalDNS` feature gate.

File: charts/seed-operatingsystemconfig/original/templates/kubelet/_kubelet-monitor.service (hunk 5 above)
Comment (timebertt) [reply charts/seed-operatingsystemconfig/original/templates/kubelet/_kubelet-monitor.service]: Ok, fair enough.

**Review by timebertt — APPROVED** [review]
Code-wise no further comments from my side.
/lgtm

**Pull request merged**